backend/session/libseat.c

```diff
@@ -0,0 +1,184 @@
+#define _POSIX_C_SOURCE 200809L
+#include <assert.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <wayland-server-core.h>
+#include <wlr/backend/session/interface.h>
+#include <wlr/config.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+
+#include <libseat.h>
+
+const struct session_impl session_libseat;
+
+struct libseat_device {
+	struct wl_list link;
+	int fd;
+	int device_id;
+};
+
+struct libseat_session {
+	struct wlr_session base;
+
+	struct libseat *seat;
+	struct wl_event_source *event;
+	struct wl_list devices;
+};
+
+static void handle_enable_seat(struct libseat *seat, void *data) {
+	struct libseat_session *session = data;
+	session->base.active = true;
+}
+
+static void handle_disable_seat(struct libseat *seat, void *data) {
+	struct libseat_session *session = data;
+	session->base.active = false;
+}
+
+static int libseat_event(int fd, uint32_t mask, void *data) {
+	struct libseat *seat = data;
+	libseat_dispatch(seat, 0);
+	return 1;
+}
+
+static struct libseat_seat_listener seat_listener = {
+	.enable_seat = handle_enable_seat,
+	.disable_seat = handle_disable_seat,
+};
+
+static struct libseat_session *libseat_session_from_session(
+		struct wlr_session *base) {
+	assert(base->impl == &session_libseat);
+	return (struct libseat_session *)base;
+}
+
+static struct wlr_session *libseat_session_create(struct wl_display *disp) {
+	struct libseat_session *session = calloc(1, sizeof(*session));
+	if (!session) {
+		wlr_log_errno(WLR_ERROR, "Allocation failed");
+		return NULL;
+	}
+
+	wl_signal_init(&session->base.session_signal);
+	wl_list_init(&session->devices);
+
+	session->seat = libseat_open_seat(&seat_listener, session);
+	if (session->seat == NULL) {
+		wlr_log_errno(WLR_ERROR, "Unable to create seat");
+		goto error;
+	}
+
+	const char *seat_name = libseat_seat_name(session->seat);
+	if (seat_name == NULL) {
+		wlr_log_errno(WLR_ERROR, "Unable to get seat info");
+		goto error;
+	}
+	snprintf(session->base.seat, sizeof(session->base.seat), "%s", seat_name);
+
+	struct wl_event_loop *event_loop = wl_display_get_event_loop(disp);
+	session->event = wl_event_loop_add_fd(event_loop, libseat_get_fd(session->seat),
```

Comment:
Minor: can we NULL-check the result here?

Reply:
Sure